File: frontend/android/app/src/main/java/com/mpz/app/KakaoLoginPlugin.java
package com.mpz.app;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.JSObject;
import com.kakao.sdk.user.UserApiClient;
import com.kakao.sdk.auth.model.OAuthToken;
import com.kakao.sdk.user.model.User;
import android.content.Context;
import android.util.Log;

@CapacitorPlugin(name = "KakaoLogin")
public class KakaoLoginPlugin extends Plugin {

    private static final String TAG = "KakaoLoginPlugin";

    @PluginMethod
    public void initialize(PluginCall call) {
        String appKey = call.getString("appKey");
        if (appKey != null) {
            Log.d(TAG, "Kakao SDK 초기화: " + appKey);
            call.resolve();
        } else {
            call.reject("appKey가 필요합니다.");
        }
    }

    @PluginMethod
    public void login(PluginCall call) {
        Log.d(TAG, "카카오 로그인 시작");
        
        UserApiClient.getInstance().loginWithKakaoAccount(getContext(), (OAuthToken token, Throwable error) -> {
            if (error != null) {
                Log.e(TAG, "카카오 로그인 실패", error);
                call.reject("카카오 계정 로그인을 완료할 수 없습니다: " + error.getMessage());
            } else if (token != null) {
                Log.d(TAG, "카카오 로그인 성공");
                JSObject ret = new JSObject();
                
                ret.put("accessToken", token.getAccessToken());
                if (token.getRefreshToken() != null) {
                    ret.put("refreshToken", token.getRefreshToken());
                }
                if (token.getIdToken() != null) {
                    ret.put("idToken", token.getIdToken());
                }
                
                call.resolve(ret);
            }
            return null;
        });
    }

    @PluginMethod
    public void logout(PluginCall call) {
        UserApiClient.getInstance().logout((Throwable error) -> {
            if (error != null) {
                Log.e(TAG, "카카오 로그아웃 실패", error);
                call.reject("로그아웃에 실패했습니다: " + error.getMessage());
            } else {
                Log.d(TAG, "카카오 로그아웃 성공");
                call.resolve();
            }
            return null;
        });
    }

    @PluginMethod
    public void unlink(PluginCall call) {
        UserApiClient.getInstance().unlink((Throwable error) -> {
            if (error != null) {
                Log.e(TAG, "카카오 회원탈퇴 실패", error);
                call.reject("회원탈퇴에 실패했습니다: " + error.getMessage());
            } else {
                Log.d(TAG, "카카오 회원탈퇴 성공");
                call.resolve();
            }
            return null;
        });
    }

    @PluginMethod
    public void getUserInfo(PluginCall call) {
        UserApiClient.getInstance().me((User user, Throwable error) -> {
            if (error != null) {
                Log.e(TAG, "사용자 정보 조회 실패", error);
                call.reject("사용자 정보를 가져올 수 없습니다: " + error.getMessage());
            } else if (user != null) {
                Log.d(TAG, "사용자 정보 조회 성공: " + user.getId());
                JSObject ret = new JSObject();
                ret.put("id", user.getId());
                
                if (user.getKakaoAccount() != null) {
                    ret.put("email", user.getKakaoAccount().getEmail());
                    
                    if (user.getKakaoAccount().getProfile() != null) {
                        ret.put("nickname", user.getKakaoAccount().getProfile().getNickname());
                        ret.put("profileImageUrl", user.getKakaoAccount().getProfile().getProfileImageUrl());
                    }
                }
                
                call.resolve(ret);
            }
            return null;
        });
    }

    @Override
    public Context getContext() {
        return bridge.getContext();
    }
}

File: frontend/ios/App/App/KakaoLoginPlugin.swift
import Foundation
import Capacitor
import KakaoSDKCommon
import KakaoSDKAuth
import KakaoSDKUser

@objc(KakaoLoginPlugin)
public class KakaoLoginPlugin: CAPPlugin {
    private var appKey: String?
    private var isInitialized = false

    @objc func initialize(_ call: CAPPluginCall) {
        let key = call.getString("appKey") ?? appKey ?? ""
        guard !key.isEmpty else {
            call.reject("Kakao appKey is missing.")
            return
        }
        KakaoSDK.initSDK(appKey: key)
        appKey = key
        isInitialized = true
        call.resolve()
    }

    private func ensureInitIfNeeded() -> Bool {
        if isInitialized { return true }
        guard let key = appKey else { return false }
        KakaoSDK.initSDK(appKey: key)
        isInitialized = true
        return true
    }

    @objc func login(_ call: CAPPluginCall) {
        guard ensureInitIfNeeded() else {
            call.reject("Kakao SDK not initialized. Call initialize(appKey) first.")
            return
        }

        let handleToken: (OAuthToken?, Error?) -> Void = { token, error in
            if let error = error {
                call.reject("Kakao login failed: \(error.localizedDescription)")
                return
            }
            guard let token = token else {
                call.reject("Kakao login failed: no token returned.")
                return
            }
            call.resolve([
                "accessToken": token.accessToken,
                "refreshToken": token.refreshToken ?? "",
                "idToken": token.idToken ?? ""
            ])
        }

        // 카카오톡 앱 우선, 미설치 시 계정 로그인
        if UserApi.isKakaoTalkLoginAvailable() {
            UserApi.shared.loginWithKakaoTalk(completion: handleToken)
        } else {
            UserApi.shared.loginWithKakaoAccount(completion: handleToken)
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        guard ensureInitIfNeeded() else {
            call.reject("Kakao SDK not initialized.")
            return
        }
        UserApi.shared.logout { error in
            if let error = error {
                call.reject("Kakao logout failed: \(error.localizedDescription)")
                return
            }
            call.resolve()
        }
    }

    @objc func unlink(_ call: CAPPluginCall) {
        guard ensureInitIfNeeded() else {
            call.reject("Kakao SDK not initialized.")
            return
        }
        UserApi.shared.unlink { error in
            if let error = error {
                call.reject("Kakao unlink failed: \(error.localizedDescription)")
                return
            }
            call.resolve()
        }
    }

    @objc func getUserInfo(_ call: CAPPluginCall) {
        guard ensureInitIfNeeded() else {
            call.reject("Kakao SDK not initialized.")
            return
        }
        UserApi.shared.me { user, error in
            if let error = error {
                call.reject("Kakao user info failed: \(error.localizedDescription)")
                return
            }
            guard let user = user else {
                call.reject("Kakao user info failed: no user returned.")
                return
            }
            call.resolve([
                "id": "\(user.id ?? 0)",
                "email": user.kakaoAccount?.email ?? "",
                "nickname": user.kakaoAccount?.profile?.nickname ?? "",
                "profileImageUrl": user.kakaoAccount?.profile?.profileImageUrl?.absoluteString ?? ""
            ])
        }
    }
}
